security/pkg/nodeagent/caclient/providers/citadel/client_test.go

```diff
@@ -107,3 +115,99 @@ func TestCitadelClient(t *testing.T) {
 		}
 	}
 }
+
+type mockTokenCAServer struct {
+	Certs []string
+	Err   error
+}
+
+func (ca *mockTokenCAServer) CreateCertificate(ctx context.Context, in *pb.IstioCertificateRequest) (*pb.IstioCertificateResponse, error) {
+	client := fake.NewSimpleClientset()
+	tokenReview := &k8sauth.TokenReview{
+		Spec: k8sauth.TokenReviewSpec{
+			Token: validToken,
+		},
+	}
+	tokenReview.Status.Audiences = []string{}
+	tokenReview.Status.Authenticated = true
+	tokenReview.Status.User = k8sauth.UserInfo{
+		Username: "system:serviceaccount:default:example-pod-sa",
+		Groups:   []string{"system:serviceaccounts"},
+	}
+	client.PrependReactor("create", "tokenreviews", func(action ktesting.Action) (bool, runtime.Object, error) {
+		return true, tokenReview, nil
+	})
+	authenticator := authenticate.NewKubeJWTAuthenticator(client, "Kubernetes", nil, "example.com", jwt.PolicyFirstParty)
+	_, err := authenticator.Authenticate(ctx)
+	if err == nil {
+		return &pb.IstioCertificateResponse{CertChain: ca.Certs}, nil
+	}
+
+	return nil, err
+}
+
+func TestCitadelClientWithDifferentTypeToken(t *testing.T) {
+	testCases := map[string]struct {
+		server       mockTokenCAServer
+		expectedCert []string
+		expectedErr  string
+		token        string
+	}{
+		"Valid Token": {
+			server:       mockTokenCAServer{Certs: fakeCert, Err: nil},
+			expectedCert: fakeCert,
+			expectedErr:  "",
+			token:        validToken,
+		},
+		"Empty Token": {
+			server:       mockTokenCAServer{Certs: nil, Err: fmt.Errorf("test failure")},
+			expectedCert: nil,
+			expectedErr:  "rpc error: code = Unknown desc = target JWT extraction error: no HTTP authorization header exists",
+			token:        "",
+		},
+		"inValid Token": {
+			server:       mockTokenCAServer{Certs: []string{}, Err: nil},
+			expectedCert: nil,
+			expectedErr:  "invalid response cert chain",
+			token:        fakeToken,
+		},
+	}
+
+	for id, tc := range testCases {
+		// create a local grpc server
+		s := grpc.NewServer()
+		defer s.Stop()
+		lis, err := net.Listen("tcp", mockServerAddress)
+		if err != nil {
+			t.Fatalf("Test case [%s]: failed to listen: %v", id, err)
+		}
+
+		go func() {
+			pb.RegisterIstioCertificateServiceServer(s, &tc.server)
+			if err := s.Serve(lis); err != nil {
+				t.Logf("Test case [%s]: failed to serve: %v", id, err)
+			}
+		}()
+
+		// The goroutine starting the server may not be ready, results in flakiness.
+		time.Sleep(1 * time.Second)
```

Comment:
The required wait time may vary depending on the platform that runs the test. Shall we add a retry to wrap the call NewCitadelClient() instead of time.Sleep here?

Reply:
ok